### PR TITLE
Introduce DxFontInfo

### PR DESCRIPTION
**Pull request by skyline75489**
This PR Introduces `DxFontInfo` to simplify the logic in
`DxFontRenderData`. 

`DxFontInfo` aims to be the DWrite equivalent of `FontInfo` &
`FontInfoBase` in GDI. It encapsulates the needed information to
represent a displayable font face. It also provides the ability to
resolve a font face based on the available fonts on the system.

## References

This is a follow-up of #9096.
Initial Italic support was introduced by #8580.

The motivation behind this is to support bold & bold-italic text in
Windows Terminal.

### CODE REVIEW

File: src/renderer/dx/DxFontRenderData.cpp
Comment (skyline75489) [inline src/renderer/dx/DxFontRenderData.cpp:77]: I feel like I'm abusing the lazy-initializing pattern but it's really handy in this specific class and really helps with the structure.

File: src/renderer/dx/DxFontRenderData.cpp
Comment (skyline75489) [inline src/renderer/dx/DxFontRenderData.cpp:125]: Note that the following lines are newly added to match the implementation of `DefaultTextFormat()`. I don't really know the implication of this, though. It seems fine without the following lines.

File: src/renderer/dx/CustomTextLayout.cpp
Comment (skyline75489) [inline src/renderer/dx/CustomTextLayout.cpp:155]: I don't know why auditing is complaining about possible exceptions even though I clearly added `CATCH_RETURN()`.

File: src/renderer/dx/CustomTextLayout.cpp
Comment (skyline75489) [inline src/renderer/dx/CustomTextLayout.cpp:139]: This is where I'll be adding `bold` and `bold-italic`. @DHowett is this a feasible plan to you? It works, of course. Just want to know if it's a good practice.

File: src/renderer/dx/DxFontInfo.h
Comment (skyline75489) [inline src/renderer/dx/DxFontInfo.h:40]: The name `SetFromEngine` is copied from `FontInfo`. I don't know why the name. This is just a "set-everything" method.

File: src/renderer/dx/DxFontInfo.h
Comment (skyline75489) [inline src/renderer/dx/DxFontInfo.h:18]: I'm not sure whether this should be `familyName` of `faceName`. Seems to me there are `familyName` & `faceName` in the GDI font info, but only ` familyName` in DWrite?

File: src/renderer/dx/DxFontRenderData.cpp
Comment (skyline75489) [inline src/renderer/dx/DxFontRenderData.cpp:428]: Might use the split mode to review this file from this point onwards.

Comment (skyline75489) [reply src/renderer/dx/DxFontRenderData.cpp:428]: This method sets `_fontSize`, `_lineSpacing`, `_lineMetrics` and `_glyphCell ` internally. I wish all of them can be explicitly acquired just like the  `DefaultTextFormat()` or `Analyzer()`, but I failed to find a way to break this method into smaller ones, because the logic inside this is tightly-coupled and I honestly understand none of it.

File: src/renderer/dx/CustomTextLayout.cpp
Comment (skyline75489) [reply src/renderer/dx/CustomTextLayout.cpp:139]: On my other branch, it looks like this:

```cpp
if (drawingContext->useBoldFont && drawingContext->useItalicFont)
{
    _formatInUse = _fontRenderData->BoldItalicTextFormat().Get();
    _fontInUse = _fontRenderData->BoldItalicFontFace().Get();
}
else if (drawingContext->useBoldFont)
{
    _formatInUse = _fontRenderData->BoldTextFormat().Get();
    _fontInUse = _fontRenderData->BoldFontFace().Get();
}
else if (drawingContext->useItalicFont)
{
    _formatInUse = _fontRenderData->ItalicTextFormat().Get();
    _fontInUse = _fontRenderData->ItalicFontFace().Get();
}
else
{
    _formatInUse = _fontRenderData->DefaultTextFormat().Get();
    _fontInUse = _fontRenderData->DefaultFontFace().Get();
}
```

File: src/renderer/dx/CustomTextLayout.cpp
Comment (miniksa) [reply src/renderer/dx/CustomTextLayout.cpp:155]: I remember having this problem before, but not what I did about it besides rip off the `noexcept` and go on with life.

**Review by miniksa — CHANGES_REQUESTED** [review]

File: src/renderer/dx/DxFontInfo.cpp
Comment (skyline75489) [inline src/renderer/dx/DxFontInfo.cpp:16]: Auditing won't let me use the other constructors because they would throw exceptions, and default constructor should not throw exceptions :/

File: src/renderer/dx/DxFontInfo.cpp
Comment (skyline75489) [reply src/renderer/dx/DxFontInfo.cpp:144]: No worries. Could you please double check that I'm not messing up anything in #9734 ?

File: src/renderer/dx/DxFontInfo.cpp
Comment (skyline75489) [inline src/renderer/dx/DxFontInfo.cpp:327]: This should be the only place where "something" changed after merging #9734. I can't use `As()` here because `dwriteFactory` is a raw pointer.

File: src/renderer/dx/DxFontInfo.h
Comment (skyline75489) [reply src/renderer/dx/DxFontInfo.h:42]: I don't quite get it what you mean by "feeding it into a fresh constructed one" . sorry my understanding of English will go offline occasionally. 

I kinda like the way it is now because it has the same name as in `FontInfo`. One day maybe we can merge `FontInfo` with `DxFontInfo`, then this name would clarify what it is.

File: src/renderer/dx/CustomTextLayout.cpp
Comment (skyline75489) [inline src/renderer/dx/CustomTextLayout.cpp:135]: See https://github.com/microsoft/terminal/pull/9201#discussion_r613463387

File: src/renderer/dx/DxFontRenderData.cpp
Comment (skyline75489) [reply src/renderer/dx/DxFontRenderData.cpp:432]: Then it's done! I can live with all the logic wrapped inside ` _BuildFontRenderData(desired, actual, dpi)` . The caller does not need to care about the details inside, and can enjoy the much clearer interface.

File: src/renderer/dx/DxFontInfo.cpp
Comment (skyline75489) [reply src/renderer/dx/DxFontInfo.cpp:144]: Also I'm glad to provide a better home for `NearbyCollection` and friends, of which the implementation detail should be hidden from the outside world.

File: src/renderer/dx/DxFontRenderData.cpp
Comment (skyline75489) [inline src/renderer/dx/DxFontRenderData.cpp:562]: Seeing both `SetFromEngine` and `SetFallback` is a bit weird but for now I can live with it. It matches the old code. 

We can probably leave it to some kids in the future.

File: src/renderer/dx/DxFontRenderData.cpp
Comment (skyline75489) [reply src/renderer/dx/DxFontRenderData.cpp:129]: are you suggesting I should remove them in both `DefaultTextFormat()` and `ItalicTextFormat()`, also future `SomeTextFormat`?

I mean I have a following PR that adds `BoldTextFormat` and friends. We can probably discuss this later.

Comment (miniksa) [reply src/renderer/dx/DxFontRenderData.cpp:129]: I'm not strictly suggesting that because I'm not 100% certain it's unused. It would take figuring that out first. For this PR... you don't really need to do anything for it and can just preserve the behavior to be safe.

File: src/renderer/dx/DxFontInfo.h
Comment (miniksa) [reply src/renderer/dx/DxFontInfo.h:42]: It's fine. You can leave it for now. I agree that perhaps the other `FontInfo` can be unified one day.

File: src/renderer/dx/DxFontInfo.cpp
Comment (miniksa) [reply src/renderer/dx/DxFontInfo.cpp:144]: I did look and it seemed fine to me.

Yes, thank you for catching that. I shouldn't have left it public.

File: src/renderer/dx/DxFontInfo.cpp
Comment (miniksa) [reply src/renderer/dx/DxFontInfo.cpp:16]: ohhhhhhhhhhhhhhhhhhhh yes ok then its fine.

File: src/renderer/dx/DxFontInfo.cpp
Comment (miniksa) [reply src/renderer/dx/DxFontInfo.cpp:327]: That's fine. Same thing underneath anyway.

File: src/renderer/dx/DxFontRenderData.cpp
Comment (miniksa) [reply src/renderer/dx/DxFontRenderData.cpp:562]: Sure

**Review by miniksa — APPROVED** [review]

**Comment by skyline75489** [comment]
Huh...the trick `[skip ci]` does not seem to work with Azure Pipelines.

See https://github.com/Microsoft/azure-pipelines-agent/issues/1270#issuecomment-486782068 . I don't honestly know what she meant by saying "Azure DevOps honors the [skip ci] commands on CI builds as the change being built has already been merged"

File: src/renderer/dx/CustomTextLayout.cpp
Comment (skyline75489) [inline src/renderer/dx/CustomTextLayout.cpp:145]: After some consideration I decided to refactor this using `std::unordered_map`, which makes it super easy to add new attributed font. This is how **bold** is done:

```cpp
if (drawingContext->useBoldFont)
{
    weight = DWRITE_FONT_WEIGHT_BOLD;
}

if (drawingContext->useItalicFont)
{
    style = DWRITE_FONT_STYLE_ITALIC;
}

_formatInUse = _fontRenderData->TextFormatWithAttribute(weight, style, stretch).Get();
_fontInUse = _fontRenderData->FontFaceWithAttribute(weight, style, stretch).Get();

```

File: src/renderer/dx/DxFontInfo.h
Comment (skyline75489) [inline src/renderer/dx/DxFontInfo.h:90]: This is from `boost::hash_combine`. I wonder if I could just bring in the original boost code here, alongside the other boost code

File: src/renderer/dx/DxFontRenderData.cpp
Comment (skyline75489) [inline src/renderer/dx/DxFontRenderData.cpp:158]: I just use this as a cache/map thing. It works fine. I wonder if `Microsoft::WRL::ComPtr` works as expected in this situation. I mean will it leak COM reference or something? I don't really know.

File: src/renderer/dx/CustomTextLayout.cpp
Comment (skyline75489) [reply src/renderer/dx/CustomTextLayout.cpp:146]: I think this is really close to @DHowett 's vision in https://github.com/microsoft/terminal/pull/8580#issuecomment-744101996, if I understand it correctly.

File: src/renderer/dx/DxFontInfo.cpp
Comment (lhecker) [inline src/renderer/dx/DxFontInfo.cpp:174]: The second time this loops around it'll not call `_FindFontFace` with the weight/stretch/style the user asked for.

File: src/renderer/dx/DxFontInfo.h
Comment (lhecker) [inline src/renderer/dx/DxFontInfo.h:80]: Normally you specialize `std::hash<DxFontInfo>` instead.

File: src/renderer/dx/DxFontInfo.h
Comment (lhecker) [inline src/renderer/dx/DxFontInfo.h:95]:
```cpp
static const auto combine = [](std::initializer_list<size_t> list) {
    size_t seed = 0;
    for (auto hash : list)
    {
        seed ^= hash + 0x9e3779b9 + (seed << 6) + (seed >> 2);
    }
    return seed;
};
```

Then do:
```cpp
return combine({ h1, h2, h3, h4, h5 });
```

Comment (lhecker) [reply src/renderer/dx/DxFontInfo.h:95]: It might also be worthwhile to comment that this is a copy of `boost::hash_combine` [like so](https://github.com/microsoft/terminal/blob/f5182355993005639a79331f2c3673485e68a90a/src/cascadia/TerminalSettingsModel/HashUtils.h#L33-L35).

Comment (skyline75489) [reply src/renderer/dx/DxFontInfo.h:95]: Ha didn’t know it was also used in Cascadia project. Makes me really want to bring the original boost code in 😅

File: src/renderer/dx/DxFontInfo.h
Comment (skyline75489) [reply src/renderer/dx/DxFontInfo.h:80]: I’m a bit against the idea of arbitrarily injecting things into ::std namespace. So I choose the other way. If you guys all favor the first way, I’m happy to change it.

File: src/renderer/dx/DxFontInfo.cpp
Comment (skyline75489) [reply src/renderer/dx/DxFontInfo.cpp:174]: I simply moved the original code here as part of this grand refactor. I didn’t modify any of these, just to make it easy to review (for Michael specifically, who wrote this in #9734)

File: src/renderer/dx/DxFontInfo.h
Comment (TBBle) [reply src/renderer/dx/DxFontInfo.h:80]: [`std::hash`](https://en.cppreference.com/w/cpp/utility/hash) specifically is an extension point at which user code is supposed to be injected via specialisation, so that hashing, hash combining, putting into `std::unordered_map` and workalikes, [`std::hash<std::variant>`](https://en.cppreference.com/w/cpp/utility/variant/hash), and any other operation that assumes `std::hash`, "just works" without needing to _also_ override the reference to `std::hash`.

Particularly, if you implement a hash function for a structure containing a `DxFontInfo`, it should look like the hash function here, combining `std::hash` of the members, rather than needing to handle `DxFontInfoHash` specially, an _assuming_ from the name that implements [`Hash`](https://en.cppreference.com/w/cpp/named_req/Hash).

**Review by PankajBhojwani — CHANGES_REQUESTED** [review]
Some of the unit tests seem to be failing, other than that this looks good to me!

**Comment by skyline75489** [comment]
OK I have no idea how & why there tests has anything to do with my code:

```log
2021-06-11T16:59:23.2143730Z Summary of Non-passing Tests:
2021-06-11T16:59:23.2144447Z     ControlUnitTests::ControlCoreTests::TestInitialize [Failed]
2021-06-11T16:59:23.2145027Z     ControlUnitTests::ControlCoreTests::TestAdjustAcrylic [Failed]
2021-06-11T16:59:23.2145574Z     ControlUnitTests::ControlInteractivityTests::TestScrollWithMouse [Failed]
2021-06-11T16:59:23.2146125Z     ControlUnitTests::ControlInteractivityTests::CreateSubsequentSelectionWithDragging [Failed]
2021-06-11T16:59:23.2146693Z     ControlUnitTests::ControlInteractivityTests::ScrollWithSelection [Failed]
2021-06-11T16:59:23.2147256Z     ControlUnitTests::ControlInteractivityTests::TestScrollWithTrackpad [Failed]
2021-06-11T16:59:23.2147786Z     ControlUnitTests::ControlInteractivityTests::TestQuickDragOnSelect [Failed]
```

**Comment by miniksa** [comment]
> OK I have no idea how & why there tests has anything to do with my code:
> 
> ```
> 2021-06-11T16:59:23.2143730Z Summary of Non-passing Tests:
> 2021-06-11T16:59:23.2144447Z     ControlUnitTests::ControlCoreTests::TestInitialize [Failed]
> 2021-06-11T16:59:23.2145027Z     ControlUnitTests::ControlCoreTests::TestAdjustAcrylic [Failed]
> 2021-06-11T16:59:23.2145574Z     ControlUnitTests::ControlInteractivityTests::TestScrollWithMouse [Failed]
> 2021-06-11T16:59:23.2146125Z     ControlUnitTests::ControlInteractivityTests::CreateSubsequentSelectionWithDragging [Failed]
> 2021-06-11T16:59:23.2146693Z     ControlUnitTests::ControlInteractivityTests::ScrollWithSelection [Failed]
> 2021-06-11T16:59:23.2147256Z     ControlUnitTests::ControlInteractivityTests::TestScrollWithTrackpad [Failed]
> 2021-06-11T16:59:23.2147786Z     ControlUnitTests::ControlInteractivityTests::TestQuickDragOnSelect [Failed]
> ```

The failed tests all look like

```
2021-06-15T11:27:11.3398459Z TAEF: A crash with exception code 0xC0000094 occurred in module "Control.Unit.Tests.dll" in process "te.processhost.exe" (pid:6860).
``` 

Where `0xC0000094` is a thrown divide by zero exception. So I'm guessing something about the setup part of those tests is initializing something that is now having a divide by zero with your change. :(

**Comment by TBBle** [comment]
The common thing in the unit tests looks like a call to `ControlCore::Initialize`, which calls `ControlCore::_updateFont` and hence `Renderer::TriggerFontChange`, which leads to `DxEngine::UpdateFont`, and hence `DxFontRenderData::UpdateFont`. This calls `DxFontRenderData::_BuildFontRenderData`, which is changed in this patch.

`DxFontRenderData::_BuildFontRenderData` has lots of division calls in it as well.

So at a guess, there's some code-path where some initialisation never happens and hence some value is 0.f. It's being hit by the unit tests' _very_ minimal initialisation, that you aren't seeing in actual testing when you run WT.

**Comment by skyline75489** [comment]
wow thanks @TBBle for the investigation. I'll get to this PR later and found out why. I promise.

File: src/cascadia/UnitTests_Control/MockControlSettings.h
Comment (TBBle) [inline src/cascadia/UnitTests_Control/MockControlSettings.h:55]: Does this change imply that we have a divide-by-zero if the user configures a font weight of 0?

Comment (skyline75489) [reply src/cascadia/UnitTests_Control/MockControlSettings.h:55]: No, the behaviour of part of the code actually changed, so it can't handle the case. I'm still working on it.

File: src/renderer/dx/DxFontInfo.h
Comment (skyline75489) [reply src/renderer/dx/DxFontInfo.h:82]: Thanks guys. I learned a lot from this thread. I'm gonna leave the choice to Captain Dustin. When he gets to this PR and make a decision, I'll be happy to use either of the above ways.

**Review by PankajBhojwani — APPROVED** [review]
Thanks!

File: src/renderer/dx/DxFontInfo.h
Comment (DHowett) [reply src/renderer/dx/DxFontInfo.h:82]: My personal preference is that we properly extend `std::hash`; it means that we can use the STL containers with less customization, and perhaps that leads to a higher likelihood that they will be deduplicated as "identical code" :smile:

**Review by DHowett — CHANGES_REQUESTED** [review]
blocking while reviewing since this has 2 s/o

File: src/renderer/dx/DxFontInfo.h
Comment (DHowett) [reply src/renderer/dx/DxFontInfo.h:97]: Let's not bring the original boost code in; i hate how much boost costs in sheer file count

File: src/renderer/dx/DxFontRenderData.cpp
Comment (DHowett) [reply src/renderer/dx/DxFontRenderData.cpp:158]: it should work fine :)

**Review by DHowett — APPROVED** [review]

**Pull request merged**
